### PR TITLE
Feature: UI for locking canvas size aspect ratio

### LINKED ISSUE #1125: Add lock button to scene width/height to keep the aspect ratio while changing one of them
image

### PR DESCRIPTION
**Pull request by ShirleyNekoDev**
closes #1125 

image
image

This PR builds upon #1171

### CODE REVIEW

**Comment by leMaik** [comment]
Wow, the UI for this is amazing! 🤩

**Comment by ShirleyNekoDev** [comment]
~ fixed wrongly copy-pasted copyright notice

**Comment by leMaik** [comment]
I'll pull this into my PR then. :thinking:

**Comment by leMaik** [comment]
Oh no, we need a rebase :scream:

**Comment by ShirleyNekoDev** [comment]
~ rebased onto master

**Comment by ShirleyNekoDev** [comment]
~ rebased onto master (and messed up git on the way)

File: chunky/src/res/style.css [hunk 1]
```diff
@@ -63,3 +63,45 @@ Hyperlink:hover {
...
+    -fx-border-style: dotted;
+    -fx-border-color: white;
+}
+.lock-toggle > .box > .mark {
+    -fx-position-shape: true;
+    -fx-scale-x: 0.45;
+
+    -fx-scale-y: 1.50;
+    -fx-background-color: #999;
+    -fx-shape: "M-2-16c-1 0-2 1-2 2l0 8c0 1 1 2 2 2l4 0c1 0 2-1 2-2l0-8c0-1-1-2-2-2zm3 2c1 0 1 1 1 1l0 6c0 1-1 1-1 1l-2 0c-1 0-1-1-1-1l0-6c0-1 1-1 1-1zM-2 4c-1 0-2 1-2 2l0 8c0 1 1 2 2 2l4 0c1 0 2-1 2-2l0-8c0-1-1-2-2-2zm3 2c1 0 1 1 1 1l0 6c0 1-1 1-1 1l-2 0c-1 0-1-1-1-1l0-6c0-1 1-1 1-1zM-1-3l2 1 0-5c0-1-1-1-1-1-1 0-1 1-1 1zM-1 2l0 5c0 1 1 1 1 1 1 0 1-1 1-1l0-4z";
```
Comment (leMaik) [inline chunky/src/res/style.css:101]: What are the sources of these icons, i.e. do we have to give some credit or did you create them?

File: lib/src/se/llbit/chunky/ui/SizeInput.java [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+  }
+
+  public int getQueuedWidthValue() {
+    return queuedWidthValueProperty().get();
+  }
+
+  public int getQueuedHeightValue() {
+    return queuedHeightValueProperty().get();
+  }
+
```
Comment (leMaik) [inline lib/src/se/llbit/chunky/ui/SizeInput.java:262]: The properties are already exposed, is there a need to expose additional getters?

```suggestion
```

File: lib/src/se/llbit/chunky/ui/SizeInput.java [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+    setSize(
+      scaleAxis(getWidthValue(), scale),
+      scaleAxis(getHeightValue(), scale)
+    );
+  }
+
+  /**
+   * Submits queued changes to the size change listener.
+   */
+  public void submitChanges() {
```
Comment (leMaik) [inline lib/src/se/llbit/chunky/ui/SizeInput.java:285]: (taste)

```suggestion
  public void applyChanges() {
```

File: lib/src/se/llbit/chunky/ui/SizeInput.java [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+    sizeChangedCallback.accept(getWidthValue(), getHeightValue());
+    changeQueuedProperty.set(false);
+  }
+
+  /**
+   * Updates the current size change listener.
+   *
+   * @param onChangeCallback if null remove current listener, otherwise add/replace current listener
+   */
+  public void addSizeChangeListener(BiConsumer<Integer, Integer> onChangeCallback) {
```
Comment (leMaik) [inline lib/src/se/llbit/chunky/ui/SizeInput.java:297]: Javadoc (and behavior) contradict the method name here :upside_down_face: 

```suggestion
  public void setSizeChangeListener(BiConsumer<Integer, Integer> onChangeCallback) {
```

File: lib/src/se/llbit/math/QuickMath.java [hunk 5]
```diff
@@ -159,4 +159,19 @@ public static float abs(float x) {
...
+  public static int gcd(int number1, int number2) {
+    if (number1 == 0 || number2 == 0) {
+      return number1 + number2;
+    } else {
+      int absNumber1 = Math.abs(number1);
+      int absNumber2 = Math.abs(number2);
+      int biggerValue = Math.max(absNumber1, absNumber2);
+      int smallerValue = Math.min(absNumber1, absNumber2);
+      return gcd(biggerValue % smallerValue, smallerValue);
+    }
```
Comment (leMaik) [inline lib/src/se/llbit/math/QuickMath.java:175]: (nit) Would a non-recursive algorithm, i.e. a loop that updates number1 and number 2, be faster?

**Review by leMaik — CHANGES_REQUESTED** [review]

File: lib/src/se/llbit/chunky/ui/SizeInput.java (hunk 2 above)
Comment (ShirleyNekoDev) [reply lib/src/se/llbit/chunky/ui/SizeInput.java]: Other JavaFX nodes do expose them for convenience reasons - now these methods have JavaDocs to give them a reason to exist 😄

File: chunky/src/res/style.css (hunk 1 above)
Comment (ShirleyNekoDev) [reply chunky/src/res/style.css:102]: self-made ~ added CC0

File: chunky/src/res/style.css [hunk 6]
```diff
@@ -63,3 +63,47 @@ Hyperlink:hover {
...
+    -fx-border-style: dotted;
+    -fx-border-color: white;
+}
+.lock-toggle > .box > .mark {
+    -fx-position-shape: true;
+    -fx-scale-x: 0.45;
+
+    -fx-scale-y: 1.50;
+    -fx-background-color: #999;
+    /* Open chains icon released under CC0 for Chunky */
```
Comment (leMaik) [inline chunky/src/res/style.css:101]:
```suggestion
    /* Open chains icon made for Chunky and released under CC0 */
```

File: chunky/src/res/style.css [hunk 7]
```diff
@@ -63,3 +63,47 @@ Hyperlink:hover {
...
+    -fx-shape: "M-2-16c-1 0-2 1-2 2l0 8c0 1 1 2 2 2l4 0c1 0 2-1 2-2l0-8c0-1-1-2-2-2zm3 2c1 0 1 1 1 1l0 6c0 1-1 1-1 1l-2 0c-1 0-1-1-1-1l0-6c0-1 1-1 1-1zM-2 4c-1 0-2 1-2 2l0 8c0 1 1 2 2 2l4 0c1 0 2-1 2-2l0-8c0-1-1-2-2-2zm3 2c1 0 1 1 1 1l0 6c0 1-1 1-1 1l-2 0c-1 0-1-1-1-1l0-6c0-1 1-1 1-1zM-1-3l2 1 0-5c0-1-1-1-1-1-1 0-1 1-1 1zM-1 2l0 5c0 1 1 1 1 1 1 0 1-1 1-1l0-4z";
+}
+.lock-toggle:selected > .box > .mark {
+    -fx-scale-y: 1.27;
+    -fx-background-color: white;
+    /* Closed chains icon released under CC0 for Chunky */
```
Comment (leMaik) [inline chunky/src/res/style.css:107]:
```suggestion
    /* Closed chains icon made for Chunky and released under CC0 */
```

**Review by leMaik — APPROVED** [review]

**Comment by leMaik** [comment]
minor bug: one can enter a 0 in the image dimensions, which causes an exception and somehow the validation thinks it's fine 

I think the ratio/resize calculation has rounding issues.
1. enter 1920 x 1080
2. apply, lock ratio
3. enter 2160 as height
→ expected: width updates accordingly, ratio 16:9
→ actual: 💥 width somehow becomes 4.330 (yes, with a comma), ratio is now 433 : 216

File: chunky/src/res/style.css [hunk 8]
```diff
@@ -63,3 +67,47 @@ Hyperlink:hover {
...
+    -fx-border-color: white;
+}
+.lock-toggle > .box > .mark {
+    -fx-position-shape: true;
+    -fx-scale-x: 0.45;
+
+    -fx-scale-y: 1.50;
+    -fx-background-color: #999;
+    /* Open chains icon made for Chunky and released under CC0 */
+    -fx-shape: "M-2-16c-1 0-2 1-2 2l0 8c0 1 1 2 2 2l4 0c1 0 2-1 2-2l0-8c0-1-1-2-2-2zm3 2c1 0 1 1 1 1l0 6c0 1-1 1-1 1l-2 0c-1 0-1-1-1-1l0-6c0-1 1-1 1-1zM-2 4c-1 0-2 1-2 2l0 8c0 1 1 2 2 2l4 0c1 0 2-1 2-2l0-8c0-1-1-2-2-2zm3 2c1 0 1 1 1 1l0 6c0 1-1 1-1 1l-2 0c-1 0-1-1-1-1l0-6c0-1 1-1 1-1zM-1-3l2 1 0-5c0-1-1-1-1-1-1 0-1 1-1 1zM-1 2l0 5c0 1 1 1 1 1 1 0 1-1 1-1l0-4z";
```
Comment (leMaik) [inline chunky/src/res/style.css:106]: Can this somehow use the `Icons.java`?

File: chunky/src/res/style.css [hunk 9]
```diff
@@ -63,3 +67,47 @@ Hyperlink:hover {
...
+}
+.lock-toggle:selected > .box > .mark {
+    -fx-scale-y: 1.27;
+    -fx-background-color: white;
+    /* Closed chains icon made for Chunky and released under CC0 */
+    -fx-shape: "M-2-14c-1 0-2 1-2 2l0 8c0 1 1 2 2 2l4 0c1 0 2-1 2-2l0-8c0-1-1-2-2-2zm3 2c1 0 1 1 1 1l0 6c0 1-1 1-1 1l-2 0c-1 0-1-1-1-1l0-6c0-1 1-1 1-1zM-2 2c-1 0-2 1-2 2l0 8c0 1 1 2 2 2l4 0c1 0 2-1 2-2l0-8c0-1-1-2-2-2zm3 2c1 0 1 1 1 1l0 6c0 1-1 1-1 1l-2 0c-1 0-1-1-1-1l0-6c0-1 1-1 1-1zM-1 5c0 1 1 1 1 1 1 0 1-1 1-1l0-10c0-1-1-1-1-1-1 0-1 1-1 1z";
```
Comment (leMaik) [inline chunky/src/res/style.css:112]: Can this somehow use the `Icons.java`?

File: lib/src/se/llbit/chunky/ui/IntegerTextField.java [hunk 10]
```diff
@@ -6,7 +6,19 @@
 public class IntegerTextField extends NumericTextField<IntegerProperty> {
 
   public IntegerTextField() {
-    super(new SimpleIntegerProperty());
-    this.getConverter().setParseIntegerOnly(true);
+    this(0);
```
Comment (leMaik) [inline lib/src/se/llbit/chunky/ui/IntegerTextField.java:9]: Does this break any other usages?

**Review by leMaik — APPROVED** [review]

File: lib/src/se/llbit/chunky/ui/IntegerTextField.java (hunk 10 above)
Comment (ShirleyNekoDev) [reply lib/src/se/llbit/chunky/ui/IntegerTextField.java:9]: You had me worried for a second, but it is the same code as before, only 4 lines lower :)

**Comment by Peregrine05** [comment]
Is this a problem for this PR?
image
As you can see, the resolution in the resolution boxes of the scene tab does not match that of the render preview.
My default resolution is set to 960x600, but the resolution boxes always display 400x400, regardless of the default or the actual setting.

**Comment by Peregrine05** [comment]
All right.

**Pull request merged**